applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py

```diff
@@ -137,6 +137,26 @@ def GetDefaultParameters(cls):
         this_defaults.AddMissingParameters(super().GetDefaultParameters())
         return this_defaults
 
+    def ValidateSettings(self):
+        """This function validates the settings of the solver
+        """
+
+        super().ValidateSettings()
+
+        # Checks if scaling is used in combination with rebuild level lower than 2 and prebuild dynamics, if so it
+        # throws an error
+        if self.settings.Has("linear_solver_settings"):
+            if self.settings["linear_solver_settings"].Has("scaling"):
+                if self.settings["linear_solver_settings"]["scaling"].GetBool():
+                    if self.settings.Has("rebuild_level"):
+                        if self.settings["rebuild_level"].GetInt() < 2:
+                            raise ValueError("Scaling can only be used if rebuild level is 2")
+                    if self.settings.Has("prebuild_dynamics"):
+                        if self.settings["prebuild_dynamics"].GetBool():
+                            raise ValueError("Scaling can not be used if prebuild dynamics is true")
```

Comment:
To avoid deep nesting I would suggest to string conditions together by using `and`:
```suggestion
        if (self.settings.Has("linear_solver_settings") and
            self.settings["linear_solver_settings"].Has("scaling") and
            self.settings["linear_solver_settings"]["scaling"].GetBool()):
            if (self.settings.Has("rebuild_level") and
                self.settings["rebuild_level"].GetInt() < 2):
                raise ValueError("Scaling can only be used if rebuild level is 2")
            if (self.settings.Has("prebuild_dynamics" and
                self.settings["prebuild_dynamics"].GetBool()):
                raise ValueError("Scaling can not be used if prebuild dynamics is true")
```

Reply:
this wont work, since it can only check self.settings["linear_solver_settings"]["scaling"].GetBool(), if it actually has "linear_solver_settings",

same for the other checks

Reply:
I think it does work. The conditions have been ordered such that they first check that an item exists (using `Has`) before attempting to access the value. If some item does not exist (i.e. `Has` returns `False`) then the other conditions are not checked (and hence that code is not executed). See the [Python reference documentation about `and`](https://docs.python.org/3/reference/expressions.html#booleans):

> The expression `x and y` first evaluates x; if x is false, its value is returned; otherwise, y is evaluated and the resulting value is returned.

Reply:
aha, thanks